textworld/generator/chaining.py

```diff
@@ -2,297 +2,524 @@
-                cpt[0] += 1
-                c.no = str(cpt[0])
-                G.add_edge(parent.no, c.no)
-                params = ", ".join(map(str, c.action.variables))
-                labels[c.no] = "{}({})".format(c.action.name, params)
-                _recur(c)
-
-        self.no = str(cpt[0])
-        labels[self.no] = repr(self.action)  # Should be None.
-        _recur(self)
-        return G, labels
-
-
-def get_failing_constraints(state):
-    fail = Proposition("fail", [])
-
-    failed_constraints = []
-    constraints = state.all_applicable_actions(data.get_constraints().values())
-    for constraint in constraints:
-        if state.is_applicable(constraint):
-            # Optimistically delay copying the state
-            copy = state.copy()
-            copy.apply(constraint)
-
-            if copy.is_fact(fail):
-                failed_constraints.append(constraint)
-
-    return failed_constraints
-
-
-def check_state(state):
-    fail = Proposition("fail", [])
-    debug = Proposition("debug", [])
-
-    constraints = state.all_applicable_actions(data.get_constraints().values())
-    for constraint in constraints:
-        if state.is_applicable(constraint):
-            # Optimistically delay copying the state
-            copy = state.copy()
-            copy.apply(constraint)
-
-            if copy.is_fact(fail):
-                return False
+        Check if an action should be allowed in this state.
 
-    return True
+        The default implementation disallows actions that would create new facts
+        that don't mention any new variables.
 
+        Args:
+            state: The current state.
+            action: The action being applied.
 
-def maybe_instantiate_variables(rule, mapping, state, max_types_counts=None):
-    types_counts = data.get_types().count(state)
+        Returns:
+            Whether that action should be allowed.
+        """
 
-    # Instantiate variables if needed
-    try:
-        for ph in rule.placeholders:
-            if mapping.get(ph) is None:
-                name = get_new(ph.type, types_counts, max_types_counts)
-                mapping[ph] = Variable(name, ph.type)
-    except NotEnoughNounsError:
-        return None
+        for prop in action.preconditions:
+            if not state.is_fact(prop):
+                if all(state.has_variable(var) for var in prop.arguments):
+                    # Don't allow creating new predicates without any new variables
+                    return False
 
-    return rule.instantiate(mapping)
+        return True
 
+    def check_new_variable(self, state: State, type: str, count: int) -> bool:
+        """
+        Check if a new variable should be allowed to be created in this state.
 
-def _assignment_sort_key(assignment):
-    rule, mapping = assignment
+        Args:
+            state: The current state.
+            type: The type of variable being created.
+            count: The total number of variables of that type.
```

Comment:
Nice. I don't think that was ever used (or even tested :-o) but it is nice that the users can use it if they want.